File: fromscratch1/src/main/scala/examples/Article.scala
package examples.Realworld

import scala.collection.mutable
import java.util.Random
import io.jvm.uuid._

object Slug {
  def apply(input: String) = slugify(input)

  def slugify(input: String): String = {
    import java.text.Normalizer
    Normalizer
      .normalize(input, Normalizer.Form.NFD)
      .replaceAll("[^\\w\\s-]", "") // Remove all non-word, non-space or non-dash characters
      .replace('-', ' ') // Replace dashes with spaces
      .trim // Trim leading/trailing whitespace (including what used to be leading/trailing dashes)
      .replaceAll(
        "\\s+",
        "-"
      ) // Replace whitespace (including newlines and repetitions) with single dashes
      .toLowerCase // Lowercase the final results
  }
}

case class Error(msg: String, code: Int)
case class Event(authToken: Option[String], body: Option[Article])

case class User(username: String, bio: String)
object User {
  var users: mutable.ArraySeq[User] = Array(User("audacioustux", "henlo meh tangim"))

  def authenticateAndGetUser(event: Event): Option[User] = {
    event.authToken.flatMap(authToken => users.find(user => user.username == authToken))
  }
}
case class Article(
    title: String,
    body: String,
    description: String,
    id: Option[String] = None,
    slug: Option[String] = None,
    createdAt: Option[Long] = None,
    updatedAt: Option[Long] = None,
    author: Option[User] = None,
    tagList: Option[Set[String]] = None,
    favoritesCount: Option[Int] = None,
    favorited: Option[Boolean] = None
)
class ArticleController {
  // var articles: mutable.Seq[Article] = mutable.Seq()

  def create(event: Event): Error | Article = {
    val authenticatedUser = User.authenticateAndGetUser(event) match {
      case Some(authenticatedUser) => authenticatedUser
      case None                    => return Error("must be logged in", 422)
    }
    var article = event.body match {
      case Some(article) => article
      case None          => return Error("article must be specified", 422)
    }

    val timestamp = System.currentTimeMillis()
    article = article.copy(
      id = Some(UUID.random.string),
      slug = Some(Slug(article.title)),
      createdAt = Some(timestamp),
      author = Some(authenticatedUser),
      favorited = Some(false),
      favoritesCount = Some(2)
    )
    // article +: articles

    article
  }
  // def reset = articles = mutable.Seq()
}

File: bloom/src/main/python/article.py
import site
from datetime import datetime
import random

from slugify import slugify

inmemorydb = {
  "users": [{ "username": "audacioustux", "bio": "a tehc enthusiast" }],
  "articles": [],
}

def authenticateAndGetUser(event):
    if inmemorydb["users"][0]["username"] == event["authtoken"]:
        return inmemorydb["users"][0]

def create(event):
    authenticatedUser = authenticateAndGetUser(event)
    if not authenticatedUser:
        return {"error": "Must be logged in.", "errorCode": 422}

    article = event["body"]["article"]
    title = slugify(article["title"])
    description = article["description"]
    body = article["body"]
    tagList = article["tagList"]

    if not article:
        return { "error": "Article must be specified.", "errorCode": 422 }
    for expectedField in ["title", "description", "body"]:
        if not article[expectedField]:
            return { "error": "${expectedField} must be specified.", "errorCode": 422 }


    id = str(abs(random.random() * pow(36, 6)));
    timestamp = datetime.now();
    slug = title + " - " + id;

    _article = {
      "id":id,
      "slug": slug,
      "title": title,
      "description": description,
      "body": body,
      "createdAt": timestamp,
      "updatedAt": timestamp,
      "author": authenticatedUser["username"],
      "tagList": set([])
    }
    if tagList:
        _article["tagList"] = set(tagList)
    

    inmemorydb["articles"].append(_article)

    _article["favorited"] = False;
    _article["favoritesCount"] = 0;
    _article["author"] = {
      "username": authenticatedUser["username"],
      "bio": authenticatedUser["bio"] or "",
      "following": False,
    };

    return _article;


def createArticle() :
  return create({
        "authtoken": "audacioustux",
        "body": {
          "article": {
            "title": "testing blabla",
            "description": "testing bloom phew phew",
            "body": "lorem *ipsum* sit amet dolor am jam kathal",
            "tagList": ["test", "test", "bloom", "lorem-ipsum"],
          },
        },
    })

def run():
  createArticle();
  inmemorydb["articles"] = [];

run
